File: 1373/1373 - PD.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    public static void main(String[] args) throws IOException {
        InputStreamReader ir = new InputStreamReader(System.in);
        BufferedReader in = new BufferedReader(ir);

        int k;
        String[] x, y;

        k = Integer.parseInt(in.readLine());

        while (k > 0) {
            x = in.readLine().split("");
            y = in.readLine().split("");

            int tam_segmento;
            int n = x.length;
            int m = y.length;
            int[][] matriz = new int[n + 1][m + 1];

            for (int i = 1; i <= n; ++i) {
                for (int j = 1; j <= m; ++j) {
                    tam_segmento = 0;
                    matriz[i][j] = Math.max(matriz[i-1][j], matriz[i][j-1]);

                    while(tam_segmento <= i-1 && tam_segmento <= j-1 && x[i-tam_segmento-1].equals(y[j-tam_segmento-1])) {
                        tam_segmento++;

                        if (tam_segmento >= k)
                            matriz[i][j] = Math.max(matriz[i][j], matriz[i-tam_segmento][j-tam_segmento]+tam_segmento);
                    }
                }
            }

            System.out.println(matriz[n][m]);

            k = Integer.parseInt(in.readLine());
        }
    }
}

File: 1373/1373 - PD.py
k = int(raw_input())

while k != 0:
    x = raw_input()
    y = raw_input()

    n = len(x)
    m = len(y)

    matriz = [[0]*(m+1) for z in xrange(n+1)]

    for i in xrange(1, n+1):
        for j in xrange(1, m+1):
            matriz[i][j] = max(matriz[i-1][j], matriz[i][j-1])
            tam_seg = 0

            while (tam_seg <= i - 1 and tam_seg <= j - 1 and x[i-tam_seg-1] == y[j-tam_seg-1]):
                tam_seg += 1

                if (tam_seg >= k):
                    matriz[i][j] = max(matriz[i][j], matriz[i-tam_seg][j-tam_seg] + tam_seg)

    print matriz[n][m]

    k = int(raw_input())